### PR TITLE
Refactor Webpack

### PR DESCRIPTION
**Pull request by lleyton**

### CODE REVIEW

File: src/renderer/modules/webpack.ts [hunk 1]
```diff
@@ -340,3 +301,101 @@ export function patchPlaintext(patches: PlaintextPatch[]): void {
...
+): RawModule | undefined;
+export function getByProps<P extends string>(
+  props: P[],
+  options: { all?: true; raw?: true },
+): RawModule[];
+export function getByProps<P extends string>(...props: P[]): ModuleExportsWithProps<P> | undefined;
+
+export function getByProps<P extends string>(
+  ...args: [P[], GetModuleOptions] | P[]
+):
```
Comment (FifiTheBulldog) [inline src/renderer/modules/webpack.ts:355]: This mess of TS makes me wince so hard but there's not much that can be done about it without separate functions like we had before

File: src/renderer/common/react.ts [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
 // @ts-expect-error - webpack exports are loaded async
 let React: typeof import("react") & {
   reactReady: Promise<void>;
   jsx: AnyFunction;
   jsxs: AnyFunction;
 } = {
-  reactReady: waitFor(
-    byPropsFilter(["__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED", "createElement"]),
+  reactReady: waitForModule(
+    byProps(["__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED", "createElement"]),
```
Comment (FifiTheBulldog) [inline src/renderer/common/react.ts:11]: I'll revisit this whole file later, for now though it'll be okay

File: src/renderer/modules/webpack.ts [hunk 3]
```diff
@@ -137,193 +144,147 @@ function getExports(m: RawModule): ModuleExports | undefined {
...
   }
 
-  return wpRequire.c[id];
-}
-
-export function getById(id: number): ModuleExports | undefined {
-  const raw = getRawById(id);
-  if (typeof raw !== "undefined") {
-    return getExports(raw);
+  if (modules instanceof Array) {
```
Comment (FifiTheBulldog) [inline src/renderer/modules/webpack.ts:209]: Let's use `Array.isArray(modules)` here instead of `instanceof`

File: src/renderer/common/react.ts [hunk 4]
```diff
@@ -1,17 +1,17 @@
...
   jsxs: AnyFunction;
 } = {
-  reactReady: waitFor(
-    byPropsFilter(["__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED", "createElement"]),
+  reactReady: waitForModule(
+    filters.props(["__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED", "createElement"]),
   ).then((react) => {
     Object.assign(React, react as typeof import("react"));
-    React.jsx = getByProps("jsx")?.jsx as AnyFunction;
+    React.jsx = (getModule(filters.props(["jsx"])) as { jsx: AnyFunction }).jsx;
```
Comment (asportnoy) [inline src/renderer/common/react.ts:14]: Any reason you can't just do
```suggestion
    React.jsx = (getModule(filters.props(["jsx"])) as AnyFunction });
```

File: src/renderer/modules/webpack.ts [hunk 5]
```diff
@@ -137,193 +144,149 @@ function getExports(m: RawModule): ModuleExports | undefined {
...
-  return (m: RawModule): m is RawModuleWithProps<P> => Boolean(getExportsForProps(m, props));
+  return typeof rawModule !== "undefined" ? getExports(rawModule) : void 0;
 }
 
-export function getByProps<P extends string>(...props: P[]): ModuleExportsWithProps<P> | undefined {
-  return byPropsInternal(props, false);
-}
+// Searcher
+
+// I'd prefer to use conditional types instead of overloading here, but I had some weird issues with it
```
Comment (asportnoy) [inline src/renderer/modules/webpack.ts:180]: Elaborate?

File: src/renderer/modules/webpack.ts [hunk 6]
```diff
@@ -137,193 +144,149 @@ function getExports(m: RawModule): ModuleExports | undefined {
...
-export function getAllByProps<P extends string>(...props: P[]): Array<ModuleExportsWithProps<P>> {
-  return byPropsInternal(props, true);
-}
+// Filters
 
-export function getAllRawByProps<P extends string>(...props: P[]): Array<RawModuleWithProps<P>> {
-  return getAllRawModules(byPropsFilter(props)) as Array<RawModuleWithProps<P>>;
-}
+export const filters = {
+  props<P extends string>(props: P[]) {
```
Comment (asportnoy) [inline src/renderer/modules/webpack.ts:219]: I'd prefer calling this `byProps`

File: src/renderer/modules/webpack.ts (hunk 5 above)
Comment (lleyton) [reply src/renderer/modules/webpack.ts:180]: TypeScript doesn't have dependent typing; https://github.com/microsoft/TypeScript/issues/33014

File: src/renderer/modules/webpack.ts (hunk 6 above)
Comment (lleyton) [reply src/renderer/modules/webpack.ts:219]: I'm unsure on this one (@FifiTheBulldog ?)

File: src/renderer/common/react.ts (hunk 4 above)
Comment (lleyton) [reply src/renderer/common/react.ts:14]: Unless I'm misreading your suggestion, that wouldn't work, as `getModule(filters.props(["jsx"]))` returns an object with a jsx key.

File: src/renderer/modules/webpack.ts (hunk 6 above)
Comment (FifiTheBulldog) [reply src/renderer/modules/webpack.ts:219]: I second @asportnoy's opinion here.

File: src/renderer/modules/webpack.ts (hunk 5 above)
Comment (asportnoy) [reply src/renderer/modules/webpack.ts:180]: Can you add that to the comment?

**Review by asportnoy — APPROVED** [review]

**Review by FifiTheBulldog — APPROVED** [review]
Thanks!

**Pull request merged**